include/jemalloc/internal/rtree_inlines.h

```diff
@@ -130,21 +130,22 @@ rtree_read(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx, uintptr_t key,
 JEMALLOC_INLINE rtree_elm_t *
 rtree_elm_acquire(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx,
     uintptr_t key, bool dependent, bool init_missing) {
-	rtree_elm_t *elm;
-
-	elm = rtree_elm_lookup(tsdn, rtree, rtree_ctx, key, dependent,
-	    init_missing);
+	rtree_elm_t *elm = rtree_elm_lookup(tsdn, rtree, rtree_ctx, key,
+	    dependent, init_missing);
 	if (!dependent && elm == NULL) {
 		return NULL;
 	}
 
-	extent_t *extent;
-	void *s;
-	do {
-		extent = rtree_elm_read(elm, false);
+	spin_t spinner = SPIN_INITIALIZER;
+	while (true) {
+		extent_t *extent = rtree_elm_read(elm, false);
 		/* The least significant bit serves as a lock. */
-		s = (void *)((uintptr_t)extent | (uintptr_t)0x1);
-	} while (atomic_cas_p(&elm->pun, (void *)extent, s));
+		void *s = (void *)((uintptr_t)extent | (uintptr_t)0x1);
+		if (!atomic_cas_p(&elm->pun, (void *)extent, s)) {
+			break;
+		}
+		spin_adaptive(&spinner);
```

Comment:
I did a simple benchmark to get an idea about the cycle count:
for (i=0; i < 128k; i++) { // 32 threads running
  CPU_SPINWAIT;
}
and each iteration takes ~20 cycles on Xeon(R) CPU E5-2660. Without the pause, each empty iter takes ~6 cycles.

That makes me wonder if the upper bound is too high in [```spin_adaptive()```](https://github.com/jemalloc/jemalloc/blob/3bd6d8e41d524737845949c44bd4961b0930965c/include/jemalloc/internal/spin_inlines.h#L16). As the loop count is (KQU(1) << spin->iteration), even iteration == 6 makes it (2^6 * 20) = ~1280 cycles. Adding the previous iterations, the total will be  ~2.5k cycles and should be sufficient to conclude a long spin. (actually probably should taking it down a bit)